### PR TITLE
Themis specification: keys and keypairs

### PR DESCRIPTION
**Pull request by ilammy**
Start describing Themis cryptography and data structure design in a series of technical specifications. Hopefully, this will make reimplementing Themis in different platforms easier as well as simplifies security audit and analysis. It may also be beneficial to new developers to serve as a map for the codebase (or rather, as a textbook which you apply to the codebase).

It will take a while to write so unfinished sections get a notice instead of content. The pages are already available so that the hyperlinks work right away.

<sub>Yeah, instead of pointlessly rewriting Themis in Rust, I started pointlessly rewriting it in English. Feel free to join.</sub>

### CODE REVIEW

**Comment by iamnotacake** [comment]
Things I've seen in this PR so far look good and well explained. Of course, one still have to understand the math behind EC & RSA to implement them properly. Or just use the existing libraries. I guess you gonna add SMessage & SSession specs later in this same PR.

File: themis/spec/common.md
Comment (storojs72) [inline themis/spec/common.md:76]: I would add that this is a Go's programming language snippet. There are other similar places:
- https://github.com/cossacklabs/product-docs/pull/25/commits/63faf3d3b0b7355f4e5b4eb0cf4ccb953cc40be3#diff-90c4f24e395ae483f38d1408d0cf47adR54
- https://github.com/cossacklabs/product-docs/pull/25/commits/1753920b0274fc8c0401660e2fae093a3e2ed245#diff-fe4809a11f4d9d1bfaa450caf9723471R130
- https://github.com/cossacklabs/product-docs/pull/25/commits/7d8fcae2e93c889cb3b7819e2d5fb4a643c021d4#diff-14ad7a0c001539a2ca8334e3e21fb1e0R231
- https://github.com/cossacklabs/product-docs/pull/25/commits/7d8fcae2e93c889cb3b7819e2d5fb4a643c021d4#diff-14ad7a0c001539a2ca8334e3e21fb1e0R355

File: themis/spec/symmetric-keys.md
Comment (storojs72) [inline themis/spec/symmetric-keys.md:75]: I would write not "weird", but "weak". But its just IMO

File: themis/spec/asymmetric-keypairs/_index.md
Comment (storojs72) [inline themis/spec/asymmetric-keypairs/_index.md:21]: I would explain here, why we use ECC keys by default. The reason is lower key sizes of ECC keys compared to RSA keys having same levels of security: https://www.globalsign.com/en/blog/elliptic-curve-cryptography (we can even add a table from that blogpost). Plus ECC crypto has lower CPU consumption (https://crypto.stackexchange.com/questions/1190/why-is-elliptic-curve-cryptography-not-widely-used-compared-to-rsa). The only advantage of RSA is that it is better established historically

File: themis/spec/asymmetric-keypairs/rsa.md
Comment (storojs72) [inline themis/spec/asymmetric-keypairs/rsa.md:343]: seems it should be "cryptographic library"

**Review by storojs72 — COMMENTED** [review]
Very valuable specification. I have just some minor suggestions. Thank you @ilammy. Looking at your PRs arrangement just requires me to do the same and dedicate more time on proper git commit and merge practice

File: themis/spec/common.md
Comment (ilammy) [reply themis/spec/common.md:76]: > I would add that this is a Go's programming language snippet

Sure. It won't hurt to be explicit.

File: themis/spec/symmetric-keys.md
Comment (ilammy) [reply themis/spec/symmetric-keys.md:75]: > I would write not "weird", but "weak" [symmetric key].

Mmm... It's not like passphrases are weak _per se_: given a sufficient length and randomness, they can be effectively as good as the keys. I guess I wanted to communicate here that passphrases have lower entropy per byte than keys, that's why they are "worse".

I'll think of a better wording. At the moment, I'm out of ideas. @vixentael?

File: themis/spec/asymmetric-keypairs/_index.md
Comment (ilammy) [reply themis/spec/asymmetric-keypairs/_index.md:21]: > I would explain here, why we use ECC keys by default.

Yeah, makes sense. I'm not sure if we should go into much detail here, but it will be good to note that ECC is preferred to RSA and we support RSA mostly for compliance reasons. Secure Message page could go in more details on what effects the key choice has on the practical performance of the underlying algorithms.

File: themis/spec/asymmetric-keypairs/rsa.md
Comment (ilammy) [reply themis/spec/asymmetric-keypairs/rsa.md:343]: > seems it should be "cryptographic library"

IDK :joy: During editing I've flipped it between "cryptography" and "cryptographic" several times. Unless there are strong opinions, I guess I'll go ahead to grep the variants in our own docs and just use the most popular one.

**Comment by ilammy** [comment]
@storojs72, thanks for review. Your opinion is very important and valuable here.

> requires me to do the same and dedicate more time on proper git commit and merge practice

Oh crap, you have uncovered my secret agenda! :scream:

**Comment by ilammy** [comment]
@iamnotacake,

> Of course, one still have to understand the math behind EC & RSA to implement them properly. Or just use the existing libraries.

I don't intend this spec as a primer on cryptographic algorithms so they won't be explained in too much detail. The preferred way is still to avoid rolling your own crypto and just use libraries maintained by experts. It's just happens so that there aren't many libraries around which can generate keys in the format Themis wants them, so we go in more detail in this case, explaining the components of the key containers and how they can be computed (but _without_ explaining how to multiply EC points or test numbers for primality).

> I guess you gonna add SMessage & SSession specs later in this same PR.

Not in this PR, but in later ones, yeah. As soon as the algorithms are understood. I'd like to keep PRs reviewable so they shouldn't contain too much content.

File: themis/spec/asymmetric-keypairs/rsa.md
Comment (ilammy) [reply themis/spec/asymmetric-keypairs/rsa.md:343]: "Cryptographic" it is.

```
$ git grep -hoi 'crypto[a-z]*' | sort | uniq -c
    351 crypto
      3 Crypto
      5 CRYPTO
      2 cryptoalgorithm
      5 cryptoalgorithms
      1 cryptoanalysis
      2 cryptographer
      3 cryptographers
    135 cryptographic
      4 Cryptographic
     12 cryptographically
     90 cryptography
     15 Cryptography
      1 cryptograpic
      1 cryptomuseum
      5 cryptoperiod
      1 cryptoperiods
     81 cryptosystem
      3 Cryptosystem
    311 cryptosystems
      5 Cryptosystems
```

File: themis/spec/common.md
Comment (ilammy) [reply themis/spec/common.md]: I've added a mention of Go before each snippet.

File: themis/spec/symmetric-keys.md
Comment (ilammy) [reply themis/spec/symmetric-keys.md]: Well, if that's about entropy then I've just written *that* as is, without calling passphrases weird. Please give it a look.

File: themis/spec/asymmetric-keypairs/_index.md
Comment (ilammy) [reply themis/spec/asymmetric-keypairs/_index.md:21]: I've added a short blurb about EC vs RSA keys, which hopefully provides a bit more context on why Themis supports multiple types of keys and which is the preferred one.

File: themis/spec/asymmetric-keypairs/rsa.md
Comment (vixentael) [reply themis/spec/asymmetric-keypairs/rsa.md]: `cryptography library` is more common used term, better for search

https://trends.google.com/trends/explore?q=cryptography%20library,cryptographic%20library
Screenshot 2020-09-11 at 14 00 15

Comment (ilammy) [reply themis/spec/asymmetric-keypairs/rsa.md:343]: > cryptography library is more common used term, better for search

Okie dokie, reverted.

**Comment by ilammy** [comment]
Are there any more blockers for this PR?

**Comment by vixentael** [comment]
> Are there any more blockers for this PR?

Please don't merge until my review approval

File: themis/spec/asymmetric-keypairs/ecc.md
Comment (vixentael) [inline themis/spec/asymmetric-keypairs/ecc.md:10]:
```suggestion
It is critical to use a "good" curve with validated security properties,
```

File: themis/spec/asymmetric-keypairs/ecc.md
Comment (vixentael) [inline themis/spec/asymmetric-keypairs/ecc.md:170]:
```suggestion
Here is a snippet in Go, illustrating how to generate EC keypair using P256 curve:
```

File: themis/spec/symmetric-keys.md
Comment (vixentael) [inline themis/spec/symmetric-keys.md:79]:
```suggestion
Secure Cell uses additional PBKDF2 rounds to preprocess passphrase data into an encryption key.
```

let's say somewhere that we use PBKDF2?

File: themis/spec/symmetric-keys.md
Comment (vixentael) [inline themis/spec/symmetric-keys.md:99]:
```suggestion
[NIST SP 800-63](https://www.nist.gov/itl/tig/projects/special-publication-800-63) suggests using 8 characters as minimum, and up to 64 characters for user-generated passphrases (see [NIST SP 800-63b, 5.1.1 Memorized Secrets](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-63b.pdf)).
```
```

File: themis/spec/symmetric-keys.md
Comment (vixentael) [inline themis/spec/symmetric-keys.md:94]:
```suggestion
As for definitely **not recommended** approaches, avoid letting users choose their cryptographic keys, as humans are rarely capable of generating cryptographically secure random sentences. 
```
I don't know where "avoid letting users choose their passphrases" is coming from. Memorizable secrets are foundation for many security controls, NIST guides have certain rules for handling memorizable secrets.

These sentences should be rewritten or removed.

File: themis/spec/asymmetric-keypairs/rsa.md
Comment (vixentael) [inline themis/spec/asymmetric-keypairs/rsa.md:348]:
```suggestion
Here is a snippet in Go, illustrating how to generate a long RSA keypair:
```

File: themis/spec/asymmetric-keypairs/ecc.md
Comment (ilammy) [reply themis/spec/asymmetric-keypairs/ecc.md:10]: Why quotes then?
```suggestion
It is critical to use a curve with validated security properties,
```

File: themis/spec/symmetric-keys.md
Comment (ilammy) [reply themis/spec/symmetric-keys.md:79]: I've explicitly avoided mentioning PBKDF2 here since it's only an overview of keys and passphrases, not Secure Cell operation. I believe, intricacies of the KDF choice belong to Secure Cell description specifically. For one, it's not only PBKDF2 there, ZRTP-style KDF is also thrown into the mix. It will be easier to have KDF description in one place so that when a new KDF is added to Secure Cell then it's not necessary to locate and edit a bunch of other places as well. Passphrase content does not really depend on the specific KDF in use.

File: themis/spec/asymmetric-keypairs/rsa.md
Comment (ilammy) [reply themis/spec/asymmetric-keypairs/rsa.md:348]: The point here isn't to have a long keypair. If you're concerned, how about we drop the word completely?
```suggestion
Here is a snippet in Go, illustrating how to generate an RSA keypair:
```

File: themis/spec/symmetric-keys.md
Comment (ilammy) [reply themis/spec/symmetric-keys.md:99]: Well... It's contestable whether guidelines for *authentication* apply to secrets used to secure data. But okay, a reference is better than no reference.

File: themis/spec/symmetric-keys.md
Comment (ilammy) [reply themis/spec/symmetric-keys.md:94]: > I don't know where "avoid letting users choose their passphrases" is coming from.

From my head, obviously.

The stress here is on *"cryptographically secure"* (i.e., with comparable security to keys). Sure, humans are choosing their passphrases and ultimately it comes down to something kept in the head. However, without training and education, typical people are unlikely to choose a strong passphrase either, just like they are usually not able to generate keys of acceptable quality from their head.

Current Secure Cell KDF parameters are tweaked for *strong* passphrases. The slowdown of KDF contributes only about a dozen additional entropy bits and the passphrase is responsible for the rest. We recommend 256-bit keys, so my overall understanding is that the passphrases should have comparable strength. Or, well, at least not expect *majority* of entropy to come from the KDF alone and allow users to choose 4-digit PINs, claiming them to be as secure as the keys generated by Themis. For that, we'd need way more KDF rounds.

Anyways, if this is controversial and we don't have any advice to give here then I'm for removing this paragraph.

File: themis/spec/symmetric-keys.md
Comment (vixentael) [reply themis/spec/symmetric-keys.md:79]: If not mention PBKD2 explicitly, maybe let's link to a page where we describe usage of PBKDF2?

File: themis/spec/asymmetric-keypairs/rsa.md
Comment (vixentael) [reply themis/spec/asymmetric-keypairs/rsa.md:348]: I don't think we should be using "good" as description of cryptographically-secure keys/keypairs in official docs, that's why I suggest other options.

File: themis/spec/symmetric-keys.md
Comment (vixentael) [reply themis/spec/symmetric-keys.md:94]: > then I'm for removing this paragraph.

or let's comment it out? maybe future-us will get back here and update this piece

File: themis/spec/symmetric-keys.md
Comment (ilammy) [reply themis/spec/symmetric-keys.md:79]: > maybe let's link to a page where we describe usage of PBKDF2?

As soon as it there. My idea was to cross-link Secure Cell spec with key/passphrase spec once it's actually ready.

File: themis/spec/symmetric-keys.md
Comment (ilammy) [reply themis/spec/symmetric-keys.md]: > or let's comment it out? maybe future-us will get back here and update this piece

I'm not a fan of leaving swathes of commented out code, or documentation. I thought about replacing it with a `<!-- TODO -->`, linking to this discussion. But after further consideration I don't really see anything of value to be put here. We already give an idea of how a passphrase can be generated and give links to NIST recommendations on the topic. I guess that should be enough. If future editors are really interested, they can `git blame` the file.

Comment (vixentael) [reply themis/spec/symmetric-keys.md]: okay please do.

File: themis/spec/symmetric-keys.md
Comment (vixentael) [reply themis/spec/symmetric-keys.md:79]: We already have mentions of PBKDF2 in docs. I suggest to put a link on Secure Cell docs. https://docs.cossacklabs.com/themis/crypto-theory/cryptosystems/secure-cell/#key-derivation-functions

**Review by vixentael — APPROVED** [review]

File: themis/spec/symmetric-keys.md
Comment (ilammy) [reply themis/spec/symmetric-keys.md]: > We already have mentions of PBKDF2 in docs. I suggest to put a link on Secure Cell docs.
> https://docs.cossacklabs.com/themis/crypto-theory/cryptosystems/secure-cell/#key-derivation-functions

Okay, I've added links to this section.

**Review by vixentael — APPROVED** [review]

**Comment by vixentael** [comment]
Merge 'em all!

**Pull request merged**
